lib/HLSL/DxilDeleteRedundantDebugValues.cpp

```diff
@@ -0,0 +1,155 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// DxilDeleteRedundantDebugValues.cpp                                        //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+//
+// Removes as many dbg.value's as possible:
+//
+// 1. Search for all scopes (and their parent scopes) that have any real (non-debug)
+//    instructions at all.
+//
+// 2. For each dbg.value, if it's refering to a variable from a scope not in
+//    the set of scopes from step 1, then delete it.
+//
+// 3. In any contiguous series of dbg.value instructions, if there are dbg.value's
+//    that point to the same variable+fragment, then delete all but the last one,
+//    since it would be the only authentic mapping for that variable fragment.
+
+#include "dxc/HLSL/DxilGenerationPass.h"
+#include "llvm/IR/DebugInfo.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+
+#include <unordered_set>
+
+using namespace llvm;
+
+namespace {
+class DxilDeleteRedundantDebugValues : public ModulePass {
+public:
+  static char ID;
+
+  explicit DxilDeleteRedundantDebugValues() : ModulePass(ID) {
+    initializeDxilDeleteRedundantDebugValuesPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override;
+};
+char DxilDeleteRedundantDebugValues::ID;
+}
+
+bool DxilDeleteRedundantDebugValues::runOnModule(Module &M) {
+  if (!llvm::hasDebugInfo(M))
+    return false;
+
+  bool Changed = false;
+
+  unsigned NumDbgInsts = 0;
+  unsigned NumNonDbgInsts = 0;
+  for (Function &F : M) {
+    for (BasicBlock &BB : F) {
+      for (Instruction &I : BB) {
+        if (isa<DbgInfoIntrinsic>(I))
+          NumDbgInsts++;
+        else
+          NumNonDbgInsts++;
+      }
+    }
+  }
+
+  // Here is an arbitrary threshold where we decide there's too many
+  // debug instructions.
+  // TODO: Issue a warning here about removing debug instructions.
+  bool TryReduceDbgInsts = NumDbgInsts >= 10*NumNonDbgInsts && NumDbgInsts >= 100000;
```

Comment:
When is it not valuable to remove redundant debug insts?   Does this make debug info worse?

Reply:
I suppose that could be worded clearer. We can remove more debug insts by ignoring the debug locations' inlinedAt locations when collecting the set of scopes. Doing that will makes it impossible to look at some variables' values in parent scopes.

Reply:
```
void foo(inout float arg) {
  arg += 10; // While on this line, you should be able to look at variables in function `bar`, unless inlinedAt is ignored.
}

void bar(float val) {
  foo(val);
}

```

Reply:
As per our offline discussion, it's actually not really necessary to do this.